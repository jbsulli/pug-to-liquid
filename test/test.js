"use strict";

/* globals it, expect */

const parseLiquid = require('../src/parse-liquid.js');

it('should parse null string', () => {
    expect(parseLiquid('', 'test.js')).toMatchSnapshot();
});

it('should parse plain text', () => {
    expect(parseLiquid('hello world!', 'test.js')).toMatchSnapshot();
});

it('should require a string', () => {
    expect(() => parseLiquid()).toThrow('Liquid must be a string');
});

it('should parse a plain liquid tag', () => {
    expect(parseLiquid('Hello, {{ first_name }}!')).toMatchSnapshot();
});

it('should parse filters', () => {
    expect(parseLiquid('Hello, {{ first_name | capitalize }}!')).toMatchSnapshot();
});

it('should parse a logic tag', () => {
    expect(parseLiquid('{% assign name = first_name | capitalize %}')).toMatchSnapshot();
});

it('should parse multiple filters', () => {
    expect(parseLiquid('{% assign name = first_name | capitalize | strip %}')).toMatchSnapshot();
});

it('should handle tag ends and filter splitters in string', () => {
    expect(parseLiquid('{% assign liquid_greeting = liquid_var | prepend:"{{ " | append:" | capitalize }}" %}')).toMatchSnapshot();
});

it('should handle string escaping', () => {
    expect(parseLiquid('{% assign name = last_name | prepend:"Hello Mr \\"" | append:"\\"" %}')).toMatchSnapshot();
});

it('should parse whitespace control characters', () => {
    expect(parseLiquid(`
        {%- if test %}
            {{- test -}}
        {%- elsif -%}
            {{- "hello " | append:"world!" -}}
        {% endif -%}
    `)).toMatchSnapshot();
});

it('should throw if tag string doesn\'t close', () => {
    expect(() => parseLiquid('{{ "world" | prepend:"hello }}')).toThrow('No string end found');
});

it('should throw if tag string doesn\'t close', () => {
    expect(() => parseLiquid('{{ "world" | prepend:"hello }}"...')).toThrow('No tag end found');
});

it('should throw if the tag is unknown', () => {
    expect(() => parseLiquid('{% unknownTag %}')).toThrow('Invalid tag');
    try {
        parseLiquid('{% unknownTag %}');
    } catch(err){
        expect(err.pos).toEqual(0);
        expect(err.line).toEqual(1);
        expect(err.col).toEqual(1);
    }
});

it('should throw when missing ending tag', () => {
    expect(() => parseLiquid('{% if test %}')).toThrow('Missing tag end');
    expect(() => parseLiquid('{% for test in testing %}{% if test %}{% endfor %}')).toThrow('Missing tag end');
});