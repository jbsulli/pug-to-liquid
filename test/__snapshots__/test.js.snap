// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle string escaping 1`] = `
Array [
  Object {
    "data": "name = last_name",
    "filters": Array [
      Object {
        "data": "\\"Hello Mr \\\\\\"\\"",
        "loc": Object {
          "end": 50,
          "filename": undefined,
          "start": 29,
        },
        "name": "prepend",
        "type": "filter",
      },
      Object {
        "data": "\\"\\\\\\"\\"",
        "loc": Object {
          "end": 64,
          "filename": undefined,
          "start": 53,
        },
        "name": "append",
        "type": "filter",
      },
    ],
    "loc": Object {
      "end": 67,
      "filename": undefined,
      "start": 0,
    },
    "type": "assign",
  },
]
`;

exports[`should handle tag ends and filter splitters in string 1`] = `
Array [
  Object {
    "data": "liquid_greeting = liquid_var",
    "filters": Array [
      Object {
        "data": "\\"{{ \\"",
        "loc": Object {
          "end": 54,
          "filename": undefined,
          "start": 41,
        },
        "name": "prepend",
        "type": "filter",
      },
      Object {
        "data": "\\" | capitalize }}\\"",
        "loc": Object {
          "end": 82,
          "filename": undefined,
          "start": 57,
        },
        "name": "append",
        "type": "filter",
      },
    ],
    "loc": Object {
      "end": 85,
      "filename": undefined,
      "start": 0,
    },
    "type": "assign",
  },
]
`;

exports[`should parse a logic tag 1`] = `
Array [
  Object {
    "data": "name = first_name",
    "filters": Array [
      Object {
        "loc": Object {
          "end": 40,
          "filename": undefined,
          "start": 30,
        },
        "name": "capitalize",
        "type": "filter",
      },
    ],
    "loc": Object {
      "end": 43,
      "filename": undefined,
      "start": 0,
    },
    "type": "assign",
  },
]
`;

exports[`should parse a plain liquid tag 1`] = `
Array [
  Object {
    "data": "Hello, ",
    "loc": Object {
      "end": 7,
      "filename": undefined,
      "start": 0,
    },
    "type": "html",
  },
  Object {
    "data": "first_name",
    "loc": Object {
      "end": 23,
      "filename": undefined,
      "start": 7,
    },
    "type": "tag",
  },
  Object {
    "data": "!",
    "loc": Object {
      "end": 24,
      "filename": undefined,
      "start": 23,
    },
    "type": "html",
  },
]
`;

exports[`should parse filters 1`] = `
Array [
  Object {
    "data": "Hello, ",
    "loc": Object {
      "end": 7,
      "filename": undefined,
      "start": 0,
    },
    "type": "html",
  },
  Object {
    "data": "first_name",
    "filters": Array [
      Object {
        "loc": Object {
          "end": 33,
          "filename": undefined,
          "start": 23,
        },
        "name": "capitalize",
        "type": "filter",
      },
    ],
    "loc": Object {
      "end": 36,
      "filename": undefined,
      "start": 7,
    },
    "type": "tag",
  },
  Object {
    "data": "!",
    "loc": Object {
      "end": 37,
      "filename": undefined,
      "start": 36,
    },
    "type": "html",
  },
]
`;

exports[`should parse multiple filters 1`] = `
Array [
  Object {
    "data": "name = first_name",
    "filters": Array [
      Object {
        "loc": Object {
          "end": 40,
          "filename": undefined,
          "start": 30,
        },
        "name": "capitalize",
        "type": "filter",
      },
      Object {
        "loc": Object {
          "end": 48,
          "filename": undefined,
          "start": 43,
        },
        "name": "strip",
        "type": "filter",
      },
    ],
    "loc": Object {
      "end": 51,
      "filename": undefined,
      "start": 0,
    },
    "type": "assign",
  },
]
`;

exports[`should parse null string 1`] = `Array []`;

exports[`should parse plain text 1`] = `
Array [
  Object {
    "data": "hello world!",
    "loc": Object {
      "end": 12,
      "filename": "test.js",
      "start": 0,
    },
    "type": "html",
  },
]
`;

exports[`should parse whitespace control characters 1`] = `
Array [
  Object {
    "data": "
        ",
    "loc": Object {
      "end": 9,
      "filename": undefined,
      "start": 0,
    },
    "type": "html",
  },
  Object {
    "data": "test",
    "loc": Object {
      "end": 23,
      "filename": undefined,
      "start": 9,
    },
    "needs_end": true,
    "trim_left": true,
    "type": "if",
  },
  Object {
    "data": "
            ",
    "loc": Object {
      "end": 36,
      "filename": undefined,
      "start": 23,
    },
    "type": "html",
  },
  Object {
    "data": "test",
    "loc": Object {
      "end": 48,
      "filename": undefined,
      "start": 36,
    },
    "trim_left": true,
    "trim_right": true,
    "type": "tag",
  },
  Object {
    "data": "
        ",
    "loc": Object {
      "end": 57,
      "filename": undefined,
      "start": 48,
    },
    "type": "html",
  },
  Object {
    "loc": Object {
      "end": 70,
      "filename": undefined,
      "start": 57,
    },
    "trim_left": true,
    "trim_right": true,
    "type": "elsif",
  },
  Object {
    "data": "
            ",
    "loc": Object {
      "end": 83,
      "filename": undefined,
      "start": 70,
    },
    "type": "html",
  },
  Object {
    "data": "\\"hello \\"",
    "filters": Array [
      Object {
        "data": "\\"world!\\"",
        "loc": Object {
          "end": 113,
          "filename": undefined,
          "start": 98,
        },
        "name": "append",
        "type": "filter",
      },
    ],
    "loc": Object {
      "end": 117,
      "filename": undefined,
      "start": 83,
    },
    "trim_left": true,
    "trim_right": true,
    "type": "tag",
  },
  Object {
    "data": "
        ",
    "loc": Object {
      "end": 126,
      "filename": undefined,
      "start": 117,
    },
    "type": "html",
  },
  Object {
    "loc": Object {
      "end": 138,
      "filename": undefined,
      "start": 126,
    },
    "trim_right": true,
    "type": "endif",
  },
  Object {
    "data": "
    ",
    "loc": Object {
      "end": 143,
      "filename": undefined,
      "start": 138,
    },
    "type": "html",
  },
]
`;
